docs(websocket-v2): add info about new websocket v2 endpoint (#5)
ultrasoundmoney/docs: builders/top-bid-websocket.md

# top bid websocket

> Provides a stream of updates of the auction's current top bid. We run several auctions and forward bids between them (see [bid-forwarding.md](bid-forwarding.md)). Two versions are available:

Reviewer: maybe "forward top bids between them"?

> * **v1** streams the top bid among the bids submitted to this auction.

Reviewer 2: "this auction" is an ambiguous frame of reference.

i don't care much for the "local" but do think we should separate the top bids streamed out from auctions running in a particular auction instance from those streamed out because they're in the view of one auction the global top bid. also it complements nicely the "global" in the next line.

how about:

```suggestion
* **v1** streams out every new local top bid: the highest bid among bids submitted directly to an auction.
```

> * **v2 (NEW)** streams the global top bid: the highest among the bids submitted to this auction and those forwarded from our other auctions.

Reviewer 2: then v2 can be:
```suggestion
* **v2** streams out the global top bid: the highest bid known to an auction among bids submitted directly and those forwarded from other auctions. note this does not mean a builder ends up outbidding their own fresher lower bids, with more stale forwarded bids. see [bid-forwarding.md](https://github.com/ultrasoundmoney/relay-docs/blob/main/builders/bid-forwarding.md) for more detail.
```

lets drop the (NEW), it rots. good for an announcement, no good for docs which should stay fresh over many months.

> We recommend v2. The global top bid is the bid the relay would deliver to the proposer at that moment, so it is the one to compete with. v2 timestamps also have nanosecond granularity, up from millisecond in v1.

Reviewer 2: ```suggestion
We recommend using v2. The global top bid is the bid the relay will deliver to the proposer at the reply deadline, so it is the one to compete with. v2 timestamps also have nanosecond granularity, up from millisecond in v1.
```

Reviewer 2: excellent note btw

Both versions send ping frames, clients should respond with pong. Use persistent connections if possible. When closing connections please make sure to close the socket properly.

> Note that each (slot, parent\_hash) combination is a separate auction with its own top bid.

Reviewer 2: ```suggestion
Note: at times, multiple parent blocks are reasonable to build on due to a split chain view. In these slots auction servers run multiple auctions. Each unique (slot, parent\_hash) combination then refers to an independent auction with its own top bid.
```

## v1

> Emits an update whenever the top bid among the bids submitted to this auction changes. Bids forwarded from our other auctions are not considered, so a bid that is winning globally but was submitted to another auction will not show up in this stream.

Reviewer 2: hmm. would be more clear if these sections describe what they do. but then up top we duplicate the info. so then better to gut the part up top so it truly only makes the recommendation and each describes itself.

messy to suggest the changes and edit all my suggestions up top. ill DM you.

}
```

You can find an example client here: [https://github.com/ultrasoundmoney/top-bid-websocket-client](https://github.com/ultrasoundmoney/top-bid-websocket-client)

## v2 (NEW)

Emits an update whenever the global top bid changes.

Endpoints (also available on the respective direct auction hosts):
* `ws://relay-builders-eu.ultrasound.money/ws/v2/top_bid`
* `ws://relay-builders-us.ultrasound.money/ws/v2/top_bid`
* `ws://relay-builders-jp.ultrasound.money/ws/v2/top_bid`

It emits SSZ encoded data of the following (rust) type:

```rust
pub struct TopBidUpdateV2 {

>     /// Nanosecond timestamp at which this became the global top bid

Reviewer: i'd change this to "...at which this became the global top bid in this auction instance". "global top bid" makes it seem like there's a single source of truth across instances. we already clarify the top bid forwarding behavior above.
